Sidecar containers KEP (#3761)
kubernetes/enhancements: keps/sig-node/3759-sidecar-containers/README.md

Some service meshes depend on secret (like a certificate) downloaded by other
Init container to establish secure communication between services. This makes
the problem of ordering of sidecar and Init containers harder.

### Problems: configuration / secrets

Some pods use init containers to pull down configuration/secrets and update them
before the main container gets it. Then use sidecars to continue to watch for
changes and perform the updates and push to the main container. This requires
two separate code paths today. Perhaps the same sidecar could be used for both
cases.

### Goals

This proposal aims to:
- make containers implementing the sidecar pattern first class citizens inside a
  Pod
- solve a Job completion issue when sidecars should run continuously
- allow mixing initContainers and sidecars for a choreographed startup sequence
- allow easy injection of sidecar containers in any Pod

### Non-Goals

This proposal doesn't aim to:
- support arbitrary dependencies graphs between containers
- act as a security control to enforce that pod containers only run while the
  sidecar is healthy. Restart of sidecar containers is a best effort

> - (alpha) support containers termination ordering

Reviewer: A relatively common problem I heard over the years was the following:

> More effectively separate access to privileged / secure resources between containers. Filesystem resources, secrets, or more privileged process behaviors can be separated into a container that has higher access and is available before the regular serving/job container and until it has completed.

This is somewhat part of the service mesh use case (especially those that perform privileged iptables manipulation and hide access to secrets), but I don't see it called out here in some form as a goal (or a really good reason it's a non goal).  I want to make sure our design does not limit or constrain such use cases.  Other examples over the years that are not service mesh:

1. Run an agent that exchanges pod identity information to gain a time or space limited security credential (such as an SSL cert, remote token, whatever) alongside a web serving container that has no access to the pod identity credential, and thus can only expose the bound credential if compromised.
2. Start a FUSE daemon in a sidecar that mounts an object storage bucket into one of the shared pod emptydir volumes - have another container running with much lower privileges that can use that
3. Run a container image build with privileges in a sidecar container that responds to commands on a unix socket shared in a pod emptydir and takes commands from an unprivileged container to snapshot the state of the container.

If this was covered in an earlier discussion, would like to see details and at least a summary here.

Author: this was covered in discussions. What we get to in these discussions, that if there will be security boundaries support like this, it will equally apply to regular Init containers and introducing sidecars does not make this (already hard) problem harder to solve.

So we said we will postpone solving it applying it to all "infrastructure" containers. I'll add a note about it.

Reviewer: I'm fine with a statement deferring it inline "we expect use cases to exist around security and we do not constrain them here (example of why we aren't constraining them)".  Just looking for enough to believe we aren't gonna be limited later based on decisions now.

## Proposal

The proposal is to introduce a `restartPolicy` field to Init containers and use
it to indicate that an initContainer is a sidecar container. The only supported
value proposed in this KEP is “Always”. Kubelet will start these containers in
the order with other Init containers, but instead of waiting for its completion,
it will wait for the container startup completion (startup probe succeeded or if

> no startup probe defined, postStart completed).

Reviewer: Also define what happens if the container exits before startup probe / postStart complete.

> no startup probe defined, postStart completed).

Author: updated. Indeed it worth spelling out

These containers will not block Pod completion - if all regular containers
complete, sidecar containers will be terminated.

Those containers will also be restarted even when the Pod restartPolicy is Never
or OnFailure. Furthermore, sidecar containers will be restarted even during Pod
termination.

In order to minimize OOM kills of sidecar containers, the OOM adjustment for
these containers will match or exceed the OOM score adjustment of regular
containers in the Pod. This intent to solve the issue
https://github.com/kubernetes/kubernetes/issues/111356 

> As part of this KEP we also will be enabling for sidecar containers: PreStart

Reviewer: Is this supposed to be a list?  Kind of hard to read, maybe use `*` for each bullet?

Author: sorry. Fixed

the Pod 'restartPolicy' flag:
- `Never` indicates that if the container returns a non-zero exit code, that pod
  initialization has failed. This value can be supported for Pods with Never and
  OnFailure restartPolicy indicating the critical Initialization step.
  Supporting this value on Pods with the the restartPolicy “Always” is not
  planned as it changes the semantic of this policy and may affect how
  deployments control the # of Pods in this deployment.
- `OnFailure` indicates the init container will be restarted until it completes
  successfully. This value may be used for Pods with the restartPolicy “Never” ,
  indicating that the Initialization step is flaky and should be retried.
- `Always` indicates that the container will be considered initialized the
  moment the startup probe indicates it is started and will be restarted if it
  exits for any reason until the pod terminates.
- (If the initContainer's restartPolicy is unset, the pod's restartPolicy is
  used to determine the Init container restart policy. For Always, OnFailure is
  used, for OnFailure - OnFailure, for Never - Never).

Since Init container and regular containers share the schema, restartPolicy

> field will be available for regular containers as well. From scenarios

Reviewer: This section is a bit confusing.  I think we need to clarify it a bit.

1. Are you intending for regular containers to have restartPolicy (as in, we have a future use case for it)?
  * If we aren't, sharing the schema is problematic and it might be better to do what ephemeral containers did and split the schemas
  * If we are, from a design perspective this sounds very confusing and I'd have a few suggestions
2. Are you intending to default the "inferred" container restart policy value based on pod policy into each container?

Reviewer: No matter what as a result of this change we need to have better API doc on init container fields, so it's probably time to split them.

Reviewer 2: IMO we should not be sharing types BUT we also make it very difficult NOT to do that.

Author: Initiated the discussion: https://github.com/kubernetes/kubernetes/issues/115362

containers on a Pod with the restartPolicy:Never, will not exactly be the
meaning of the restartPolicy:Always for the Pod with the restartPolicy:Always.
For the first case, containers will not block the Pod termination. For the
second, they will. In other words, the container with restartPolicy:Always will
effectively become a sidecar container similar to Init container with the same
field value.

Additional considerations can be found in the older document that was a previous
attempt to implement the sidecar containers:
https://docs.google.com/document/d/1gX_SOZXCNMcIe9d8CkekiT0GU9htH_my8OB7HSm3fM8/edit#heading=h.71o8fcesvgba.

### Naming

#### Collection name

For this KEP it is important to have sidecar containers be defined among other
Init containers to be able to express the initialization order of containers.
The name `initContainers` is not a good fit for sidecar containers as they
typically do more than initialization. The better name would be “infrastructure”
containers. The current idea is to implement sidecars as a part of init
containers and if this introduces too much trouble, the new collection name may

> replace the old collection name in future. 

Reviewer: The advantages of keeping them in initContainers are:

* The name initContainers implies something outside of normal operation of the pod
* More separate sections can lead to less clarity (users have to understand how they relate)
* initContainers have order, and we can reuse the ordering semantics to say that sidecar containers are started in FIFO order, and shut down in LIFO order, and if we want to (in the future) come up with parallel shutdown, we can do so without breaking existing users.

Author: the idea here is to replace the entire collection - deprecate Init and replace it with Infrastructure.

Author: polished text a little bit

#### Reuse of restartPolicy field and enum

The per-container restart policy was a long standing request from the community.
Implementing per-container restart policy introduces a set of challenging
problems for the pod lifecycle. For example, the state keeping for containers
which has already run to completion. Introducing sidecar containers is another
scenario where this field can be semantically used.

Introducing this field on containers opens up opportunities to implement those
long-standing requests from the community.

#### Use “Always” vs. New enum value

The semantic of an “Always” enum value is very close to what sidecars require.
This is why reusing Always to represent sidecars makes a lot of sense for Init
containers.

Why “Always” that may not look similar to Always for other containers makes
sense:

Pros:
- Allows the same enum values to be used for the `restartPolicy` of both pods
and containers, but semantics that are mostly the same. Cons:

> - There are slight differences between the semantics of `Always` for containers

Reviewer: Since these are two different fields (container restartPolicy and pod restartPolicy) with very similar semantic meaning (impacting behavior of containers), AND pod restartPolicy influences default container restartPolicy, I don't think this is a huge issue, and agree with your last comment.

and `Always` for pods. The main difference is that a initContainer with `Always`
will be terminated when the pod terminates and has no influence over the pod
lifecycle. Also, for Pods, the default `restartPolicy` is
[documented](https://github.com/kubernetes/kubernetes/blob/280473ebc4e45f9001f5f9789c318ff7329bc5f0/staging/src/k8s.io/api/core/v1/types.go#L2753-L2764)
as `Always` but for initContainers it will default to `OnFailure`. We believe
this can be addressed in the documentation of the existing enum fields.

> When in future we will support “Always” on regular containers, there will be

Reviewer: I am a bit confused - I think a regular Kube user reading this doc would think "a regular container's restart policy is *already* Always".  An init container's restart policy is "OnFailure" for restart always pods, but there is no possible "Always" for init containers today.

I think a slightly different way to discuss this is:

> Will a user be confused that "restartPolicy: Always" on an init container means "run the container more than once in a row even if it succeeds"?

And I think that is likely, because no init container runs more than once today if successful.  I think reducing that confusing is probably important, and that would lead me for that reason alone to suggest another constant value (such as `WithPod` or `UntilPodShutdown`)

Author: this confusion is highlighted in the document and the agreement is to proceed with the `Always` while keep asking for additional feedback

interesting case of “Always” having a meaning of non-blocking container for the
Pod with restartPolicy “Never” or “OnFailure”. This is easy to explain - Pod
lifecycle is controlled by containers with the matching restartPolicy. But it
may be slightly confusing.

  
#### Alternatives

> - Introduce a new collection `infrastructureContainers` that replaces

Reviewer: I don't like this because it removes the possibility of ordering init containers and infra containers, and since there ARE use cases where infra containers need init containers, I don't want to have cross ordering, so that is a very strong argument to me that we should keep them both as types of init containers.

Author: I reworded. This alternatvie was about renaming the entire `initContainers` section in favor of a new name. Alternative with the sidecar containers as a separate collection is listed in the end of the document as rejected

  semantically `initContainers`. This collection will allow both - Init
  containers and sidecar containers. Decision on this alternative can be
  postponed to after alpha implementation.
- Introduce a new enum value `UntilPodTermination` for `restartPolicy` on
  containers. This alternative was rejected because the “Always” semantic is

>   quite easy to explain and potential confusion seems to be minimal.

Reviewer: I don't think I agree, and since we can avoid the possibility of confusing with a net new feature, we should.  I would recommend we draft two different godocs and scenarios and test user feedback with them.

Author: this was added to the document

> - Introduce a new field: “ambient: true” on all containers. This property will

Reviewer: This seems to be assuming that the structs of the two types of containers have to be the same, which I don't think is true.  But I still like this option less than restartPolicy.

Author: point taken. And yes, I assume that we do not fork container definition. We can if needed

understand potential side effects from implementing the feature. We are looking
for side effects like abuse of the functionality to implement something
error-prone that sidecars were not designed for. Or causing issues like noisy
neighbors, etc.

### Scenario 1. User decides to use sidecars as a way to run regular containers

At least one regular container is required by the Pod. Users may decide to run
all other containers as sidecars and have a single container with the Sleep
logic inside with timeout or waiting for some cancellation signal from custom
job orchestration portal. This way users may implement execution with a time
limit defined by that logic. This is something that users cannot express today
“declaratively” and will be possible using new sidecar containers. One may
imagine a scenario when a third party jobs orchestration tool converts all
containers to sidecar containers programmatically and injects a single job
orchestration container into the Pod. This can be seen as an abuse of the
concept, and may lead to issues when pods terminate unexpectedly or Jobs meant
to be run as regular containers start being run as sidecars being restarted on

> Completion. However we do not see a major problem with it

Reviewer: I would suggest completing this sentence with a reason why you don't see a problem - "users can already abuse init containers, and it's not an issue" or whatever the best reason was.

Author: finished the sentence

Another reason for users to implement regular containers as sidecars would be to
use any special properties kubelet may apply to sidecar containers. For example,
if restart backoff timeouts will be minimized for sidecar containers (which was
proposed and rejected), users may decide to use this feature for regular
containers by running those as sidecars. Current proposal as written doesn’t

> introduce any special properties for sidecars that users may start abusing.

Reviewer: This feels like a stretch, which is fine.

Author: yep. We listed it here as a reference if we will decide to implement more features for sidecar containers that will not be supported by regular containers

### Scenario 2. Balloon sidecars

Users may decide to start the “large request” sidecar containers early to
pre-allocate resources for other containers. The same time asking for less
resources for a regular containers hoping to reuse what is allocated by the
“baloon” sidecar. This was previously impossible to pre-allocate resources like

> CPU before any Init containers run. If this pattern have any benefits, people

Reviewer: Technically the pod cgroup is max(highest(init_container), sum(container)), so a burst init container could balloon anyway today.

Author: there is a resource reuse logic when scheduling regular containers that potentially can not schedule regular containers to the same place. This is especially for resource managers like CPU manager or topology manager. This balooning is locking resources early. Not a big deal though.

Author: added a sentence about resource managers

> # KEP-3759: Sidecar containers

Author: using an issue owned by somebody else is hard - editing of issue is not possible to align with the current status.

- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Sidecar containers are a new type of containers that start among the Init
containers, run through the lifecycle of the Pod and don’t block pod

> termination. Kubelet makes a best effort to keep it alive and running while

Reviewer 3: keep them alive

Author: ```suggestion
termination. Kubelet makes a best effort to keep them alive and running while
```

fork of Kubernetes with this sidecar functionality added (not all
implementations are the same, but more than one company has a fork for this).

### Problems: jobs with sidecar containers

Imagine you have a Job with two containers: one which does the main processing
of the job and the other is just a sidecar facilitating it. This sidecar could
be a service mesh, a metrics gathering statsd server, etc.

When the main processing finishes, the pod won't terminate until the sidecar
container finishes too. This is problematic for sidecar containers that run
continuously.

There is no simple way to handle this on Kubernetes today. There are
work-arounds for this problem, most of them consist of some form of coupling
between the containers to add some logic where a container that finishes
communicates it so other containers can react. But it gets tricky when you have
more than one sidecar container or want to auto-inject sidecars.

The sidecar will also not be restarted for jobs with `restartPolicy:Never`
when it was OOM killed, which may render the pod unusable if the sidecar provided secure communication to other containers. The issue gets complicated by the fact that 

> sidecar containers typically has smaller request, making them the first target

Reviewer 3: have smaller

Author: ```suggestion
sidecar containers typically have smaller request, making them the first target
```

### Problems: configuration / secrets

Some pods use init containers to pull down configuration/secrets and update them
before the main container gets it. Then use sidecars to continue to watch for
changes and perform the updates and push to the main container. This requires
two separate code paths today. Perhaps the same sidecar could be used for both
cases.

### Goals

This proposal aims to:
- make containers implementing the sidecar pattern first class citizens inside a
  Pod
- solve a Job completion issue when sidecars should run continuously
- allow mixing initContainers and sidecars for a choreographed startup sequence
- allow easy injection of sidecar containers in any Pod

### Non-Goals

This proposal doesn't aim to:
- support arbitrary dependencies graphs between containers
- act as a security control to enforce that pod containers only run while the
  sidecar is healthy. Restart of sidecar containers is a best effort
- allow to enforce security boundaries to sidecar containers different that
  other containers. For example, allow Istio to run privileged to configure ip
  tables and disable other containers from doing so

> - (alpha) support containers termination ordering

Reviewer 4: This doesn't make sense to me. Either we do have support for ordering container termination, or we don't. If it's a goal but we don't mind that ab alpha doesn't include it yet, we should list it as a goal and then also mention the relevant details under Graduation Criteria.

Author: @jpbetz what do you think of stating the goal: "Allow to implement sidecar containers that will guaranteed to be running longer than regular containers"? I don't think ordering is something we are ready to commit on.

Reviewer 2: I think we have a goal that sidecar containers are guaranteed to shutdown strictly after app containers.  I strongly suspect that this KEP will require more than that, but it's still a suspicion rather than a goal.  It's not a non-goal.

I'd call it out under goals as "to be evaluated and decided before beta"

Reviewer 3: Agree that we should include the goal of shutting sidecars after apps :+1:

Author: added

## Proposal

The proposal is to introduce a `restartPolicy` field to Init containers and use
it to indicate that an initContainer is a sidecar container. The only supported

> value proposed in this KEP is `Always`. Kubelet will start these containers in

Reviewer 4: I agree with https://github.com/kubernetes/enhancements/pull/3761/files#r1084703693 - we should spell out the meaning if the user does not specify a value.

Author: updated

likely wait for sidecar initialization, effectively slowing down Pod startup
comparing to current approach. This can be eliminated by slight redesign of a
sidecars.

## Design Details

### Backward compatibility

The new field means that any Pod that is not using this field will behave the
same way as before.

The new field will only work with the proper control plane and kubelet. Upgrade
and downgrade scenarios will be covered in the further sections.

Outside of Kubernetes-controlled code, there might be third party controllers or
existing containers relying on the current behavior of Init containers.
Behaviors they can rely on:

- Assuming nothing is running in the Pod when Init container starts. For
  instance taking PID of the process assuming there are no sidecars.
- Incorrectly calculating the resource usage of a Pod not taking the new type of
  containers into account (e.g. some grafana dashboards may need modification)
- Stripping the restartPolicy:Always from Init container as an unknown field
  rendering Pod unfunctional as it will not pass the Init stage after this
- OPA rules may reject the Pod with new unknown fields because of failure to
  parse the new field

> These potential incompatibilities will be documented.

Reviewer 4: This would break (some) existing code, right? Do our API versioning rules allow for this?

Author: Discussed with @thockin and at PRR meeting. General feedback - this is inevitable and there is no way around it except documenting the behavior. We analyzed risks with people blindly porting existing sidecar container to this new model in the section "Risks and Mitigations" and didn't find anything that will likely break badly.

Reviewer 4: > this is inevitable and there is no way around it except documenting the behavior

OK, but that's answering a different question.

It sounds like we think out API guarantees _do_ allow this - which is also OK, but it means our documentation in that area probably needs to be more clear so that people implementing APIs need to know the kinds of assumptions they can't safely make.

The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

- `<package>`: `<date>` - `<test coverage>`

##### Integration tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- <test>: <link to test coverage>

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

- <test>: <link to test coverage>

### Graduation Criteria

#### Alpha

- Feature implemented behind a feature flag

> - Initial e2e tests completed and enabled

Author: added

sidecars.

## Production Readiness Review Questionnaire

<!--

Production readiness reviews are intended to ensure that features merging into
Kubernetes are observable, scalable and supportable; can be safely operated in
production environments, and can be disabled or rolled back in the event they
cause increased failures in production. See more in the PRR KEP at
https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.

The production readiness review questionnaire must be completed and approved
for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

Feature gate (also fill in values in kep.yaml)
- Feature gate name: `SidecarContainers`

> - Components depending on the feature gate: kubelet, kube-apiserver

Reviewer 4: I think this belongs under  “**How can this feature be enabled / disabled in a live cluster?**”

Author: fixed

> behaviors can be separated into a container that has higher access and is
> available before the regular serving/job container and until it has completed.

This is somewhat part of the service mesh use case (especially those that
perform privileged iptables manipulation and hide access to secrets).

Other scenarios:

- Run an agent that exchanges pod identity information to gain a time or space
  limited security credential (such as an SSL cert, remote token, whatever)
  alongside a web serving container that has no access to the pod identity
  credential, and thus can only expose the bound credential if compromised.
- Start a FUSE daemon in a sidecar that mounts an object storage bucket into one
  of the shared pod emptydir volumes - have another container running with much
  lower privileges that can use that
- Run a container image build with privileges in a sidecar container that
  responds to commands on a unix socket shared in a pod emptydir and takes
  commands from an unprivileged container to snapshot the state of the
  container.

However this alternative was rejected as the change introduced is beyond the

> scope of the sidecar KEP and will be needed fro many more scenarios.

Reviewer 5: ```suggestion
scope of the sidecar KEP and will be needed for many more scenarios.
```

Implementing sidecar containers as proposed in this KEP does not make this

> alternative harder tro implement in future.

Reviewer 5: ```suggestion
alternative harder to implement in future.
```

implementations are the same, but more than one company has a fork for this).

### Problems: jobs with sidecar containers

Imagine you have a Job with two containers: one which does the main processing
of the job and the other is just a sidecar facilitating it. This sidecar could
be a service mesh, a metrics gathering statsd server, etc.

When the main processing finishes, the pod won't terminate until the sidecar
container finishes too. This is problematic for sidecar containers that run
continuously.

There is no simple way to handle this on Kubernetes today. There are
work-arounds for this problem, most of them consist of some form of coupling
between the containers to add some logic where a container that finishes
communicates it so other containers can react. But it gets tricky when you have
more than one sidecar container or want to auto-inject sidecars.

The sidecar will also not be restarted for jobs with `restartPolicy:Never`
when it was OOM killed, which may render the pod unusable if the sidecar provided secure communication to other containers. The issue gets complicated by the fact that 
sidecar containers typically have smaller request, making them the first target

> as OOM score adjustment uses the request as a main input for calculation.

Reviewer 3: Open question: Should OOM score be set differently for these containers?

Author: differently than what? This sentence is saying that OOM score adjustment will match the Pod's OOM score adjustment.

### Problems: logging/metrics sidecar

> A logging sidecar should start before several other containers, to simplify

Reviewer 3: I think we should rewrite this use caseto mention a `log forwarding` instead of a `logging` sidecar. Logging is handled natively by container runtimes today and doesn't have races mentioned here.

Author: done

If the logging container starts after the main app, special logic needs to be
implemented to gather logs from the main app. Furthermore, if the logging
container is not yet started and the main app crashes on startup, those logs are
more likely to be lost (depends if logs go to a shared volume or over the
network on localhost, etc.). While you can modify your application to handle
this scenario during startup (as it is probably the change you need to do to
handle sidecar crashes), for shutdown this approach won't work.

On shutdown the ordering behavior is arguably more important: if the logging
container is stopped first, logs for other containers are lost. No matter if
those containers queue them and retry to send them to the logging container, or
if they are persisted to a shared volume. The logging container is already
killed and will not be restarted, as the pod is shutting down. In these cases,
logs are lost.

The same things regarding startup and shutdown apply for a metrics container.

### Problems: service mesh

Service mesh presents a similar problem: you want the service mesh container to
be running and ready before other containers start, so that any inbound/outbound

> connections that a container can initiate goes through the service mesh.

Reviewer 3: any containers in the pod initiate go through..

Author: ```suggestion
connections that a container can initiate go through the service mesh.
```

A similar problem happens for shutdown: if the service mesh container is
terminated prior to the other containers, outgoing traffic from other apps will
be blackholed or not use the service mesh.

> However, as none of these are possible to guarantee, most service meshes (like

Reviewer 3: `need to do hacks` -> `use workarounds` and maybe link to them?

Author: done

value proposed in this KEP is `Always`. Kubelet will start these containers in
the order with other Init containers, but instead of waiting for its completion,
it will wait for the container startup completion (startup probe succeeded or if
no startup probe defined, `postStart` handler completed).

Other values for `restartPolicy` of containers will not be accepted and
containers will follow the logic is currently implemented (documented
[here](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#:~:text=if%20a%20pod's%20init%20container%20fails%2C%20the%20kubelet%20repeatedly%20restarts%20that%20init%20container%20until%20it%20succeeds.%20however%2C%20if%20the%20pod%20has%20a%20restartpolicy%20of%20never%2C%20and%20an%20init%20container%20fails%20during%20startup%20of%20that%20pod%2C%20kubernetes%20treats%20the%20overall%20pod%20as%20failed.)
and more details can be found in the next section "Future use of `restartPolicy`
field").

Sidecar containers will not block Pod completion - if all regular containers
complete, sidecar containers will be terminated.

During the sidecar startup stage it will behave similarly to Init containers. If

> Pod `restartPolicy` is `Never`, sidecar container failed during startup will NOT

Reviewer 3: If restartPolicy is Never it is not a side car, right?

Author: This refers to Pod `restartPolicy`

be restarted and the whole Pod will fail. If Pod `restartPolicy` is `Always` or
`OnFailure`, it will be restarted.

Once sidecar container is started (postStart completed and startup probe
succeeded), this containers will be restarted even when the Pod `restartPolicy`
is `Never` or `OnFailure`. Furthermore, sidecar containers will be restarted
even during Pod termination.

In order to minimize OOM kills of sidecar containers, the OOM adjustment for
these containers will match or exceed the OOM score adjustment of regular

> containers in the Pod. This intent to solve the issue

Reviewer 3: okay, this covers my earlier comment above.

https://github.com/kubernetes/kubernetes/issues/111356 

> As part of this KEP we also will be enabling for sidecar containers:

Reviewer 2: Will thes apply only to sidecars or to all init containers?

IOW, if I have a "classic" initContainer, can I add a postStart hook?  If not, please spell it out here.  If so, please make clearer.

- PreStart and PostStop lifecycle handlers for sidecar containers
- All probes (startup, readiness, liveness)
- Readiness probes of sidecars will be used to determine the whole Pod
  readiness, but only after the main containers initialization started.
  Readiness probes will not start running before the Pod initialization stage
  completed 

> Since readiness probes will not start for sidecar containers till after

Reviewer 2: This goes to my above comment - it says readiness probes don't start but doesn't explain WHY that decision is correct.

Author: Mostly to make sure that all readiness probes will start at the same time. Also, there is a strange notion of sidecar becoming not ready while other Init containers are running. Will this indicate that sidecar needs to restart? Or that initialization stage keeps running?

Splitting startup and readiness eliminates these questions

Reviewer 2: I don't know if I buy it.  Readiness is there for a container to indicate that it is consumable.  Moving ahead without waiting for that feels wrong.

> Also, there is a strange notion of sidecar becoming not ready while other Init containers are running. Will this indicate that sidecar needs to restart?

Readiness does not cause restarts.

> Splitting startup and readiness eliminates these questions

You are, in effect saying "if you care about making sure your sidecar is ready, you have to use a `startupProbe`".  Does that feel weird to you?

Author: I'm getting in my head thru the scenario when sidecar wants to indicate that necessary initialization complete and other initialization can be started, but we are not ready to mark the pod as ready as we are doing some final initialization steps. Maybe even pinging from the sidecar other containers to be ready.

@howardjohn wdyt from the scenarios perspective?

Reviewer 2: Pod readiness is only barely relevant here - a container can be ready without the pod being ready.  This is a discussion of a single container.

Reviewer 6: My expectation was that once startup passes, the other containers could initialize and readiness probes would start (So readiness probes are the same - always immediately after startup passes).

But I hadn't considered the "no startup probe case"; given these are brand new usages of Pod we could say its required in order to have any of the startup blocking order.

From my POV, I don't care too much what probe signals it, so whatever is most consistent for K8s seems reasonable

Author: yes, I think it will be easier to just follow the existing pattern with startup and readiness probes. I now remember where this came from. From Risks and Mitigations section:

> Another potential problem may be that sidecars will wait for some condition to
mark itself “started” that cannot be met with the new pattern. For example, wait
for other containers to fully start. As sidecar will not succeed startup probes,
other Init containers will not run and Pod will be stuck on initialization. For
example, Knative's sidecar does aggressive probing of the user's container to
ensure they're ready prior to marking the sidecar ready itself (not for startup
probes though). This prevents the Pod from being included K8s ready endpoints.
However this problem will be discovered right away during the testing stage and
changes will need to be made to sidecar implementation.

Author: this is related to existing sidecars that just blindly ported to the new model. I am not sure how important this scenario is, but I also don't see any issues with splitting probes.

Reviewer 5: @SergeyKanzhelev what do you mean with splitting probes exactly?

I don't see why the knative scenario wouldn't work if their sidecar reports as ready when it is ready. Then, when the main container is also ready, the pod will be added to the endpoints. 

Therefore, I don't see why sidecars won't wait on _readiness_ to continue.

Am I missing something?

Author: imagine you have a sidecar that looks at other containers readiness before marking itself ready. If kubelet will wait for ready signal (instead of "started"), than the container will never get ready - no containers will be started before sidecar initialization complete.

Splitting probes means that once startup probe succeeded, other Init containers initialization will happen and eventually regular containers will be started. Readiness probe for sidecar will start at later stage and kubelet will only look for ready signal to mark Pod as ready and do things with Endpoints. But not for the lifecycle control.

Reviewer 2: > imagine you have a sidecar that looks at other containers readiness before marking itself ready.

This is challenging.  There *are* use-cases out there where one container serves readiness for another container.  For example, suppose your main app's real mechanism for detecting readiness is only exposed on  localhost, and you don't control it (e.g. an image you got from a vendor).  You can run a sidecar which does receives kubelet's probes, converts those to localhost, and serves the results.  In that case I would still configure it as the readiness probe OF THE "real" container, but to a port which just happens to be served by the helper.  The helper may have its own readiness.

So is there really a reason for one container's readiness to depend on a different container's readiness?  This feels like a gross violation of the principle of readiness.  @evankanderson - can you speak to this?  Does knative really do this?

Author: I'm not defending the pattern and received this as a point against readiness during one of the meetings.

If this pattern exists - @thockin what issues do you see with making "started" decision separate from "ready"? Readiness probes semantically are not about lifecycle, more about connectivity. Startup and liveness are about lifecycle and looks like a better fit here.

Reviewer 2: I don't think it's the worst thing to depend on starupProbe for this, but it's "non-obvious", so I would like to be sure that a) this is really what we want; and b) that we document it clearly for this use-case.  For example, startupProbs DO NOT block subsequent "plain" containers from starting.  We know `initContainers` are different - let's be clear about how.

Reviewer 7: Knative implements a "pass-through" of the original pod readiness to implement sub-second readiness probing (I believe a KEP was raised, but there were concerns about kubelet load in the general case).

Particularly for small, quick-starting containers like go binaries, we found increasing the tempo of readiness checks provided a 30%-ish reduction in cold start times.

It's disgusting, but we replace the container's declared probes with a probe of the sidecar, and pass the original container's probes to the sidecar as an argument of some type (I forget which type).

Reviewer 7: More generally, one might imagine that (even without the dependency on another pod), a sidecar might want to start up, indicate that further processing can begin, but continue loading state and indicating that the Pod as a whole is not _ready_ until the state has finished loading.

(For example, this might include Istio starting up, installing the correct firewall rules in the Pod, and then marking `startup done` while it loads the mesh state. If subsequent init containers attempt to access the network, they will be blocked until the state is loaded, but if the init containers do _something else_, then they can proceed in parallel.)

Reviewer 7: > `startupProbe`s DO NOT block subsequent "plain" containers from starting

Is there a notion of "subsequent 'plain' containers"? I'd always viewed the ordering of initContainers as a special thing.

Reviewer 2: > Particularly for small, quick-starting containers like go binaries, we found increasing the tempo of initial readiness checks provided a 30%-ish reduction in cold start times.

This is, in part, why we added startupProbes.

> It's disgusting

Your words, not mine :)

> but we replace the container's declared probes with a probe of the sidecar

I can interpret this 2 ways, which one is right:

1) The readinessProbe field of the user-defined container is changed to point to a port on the sidecar.
2) The readinessProbe of the sidecar container acts as a proxy to the user-defined container's port. 

> a sidecar might want to start up, indicate that further processing can begin, but continue loading state and indicating that the Pod as a whole is not ready until the state has finished loading

This is awkward - I'm ready enough to use but not really ready enough to USE use.  I get the Istio case.  It's not even that unbelievable, it's just ... clunky.  I'd put the traffic capture into a synchronous init container, then the sidecar into a non-blocking one, and "ready" means "can accept connections" not "have all state needed".

That said, how hypopthetical is that case?  Maybe  @howardjohn has an opinion?

Reviewer 2: > Is there a notion of "subsequent 'plain' containers"? I'd always viewed the ordering of initContainers as a special thing.

We don't guarantee sequential start, but it is sequential and I believe in Hyrum's law...

Author: > > `startupProbe`s DO NOT block subsequent "plain" containers from starting
> 
> Is there a notion of "subsequent 'plain' containers"? I'd always viewed the ordering of initContainers as a special thing.

I think this is referring to the fact that regular containers are not blocked on startupProbe execution. `postStart` lifecycle handlers are blocking and executed sequentially for regular containers while the startup probes for regular containers are all running in parallel. So there is a sequencing for regular containers, it just doesn't depend on startup probes.

Thinking about it - maybe just waiting for `postStart` to complete can be enough for sidecars. This will make containers behave similarly and will not require the probes loop to start very fast so startup will not be slowed down by it. Wdyt?

Reviewer 6: The Istio case (in https://github.com/kubernetes/enhancements/pull/3761#discussion_r1089485312) is, at least for Istio, not something we would ever do. That is already the current state (just split between init container and app container) and its broken because we don't want to expose user's containers to a broken network.

Author: I feel that making startup to only wait for `postStart` solves all the scenarios and in fact easier to implement. Any onbjections on this thread? I am rewriting this commit...

Reviewer 6: `postStart` blocking for readiness has some debugging challenges in how logs are reported (they aren't) and the status is displayed (I forget the details, but it doesn't signal "the pod is waiting", it looks pretty obscure IIRC). Perhaps we improve that situation, but it doesn't feel "right" to me. That being said, it meets our use cases so I am fine with it if others are

Reviewer 2: @evankanderson I didn't get an answer to this question:

>> but we replace the container's declared probes with a probe of the sidecar
>
> I can interpret this 2 ways, which one is right:
> 
> 1) The readinessProbe field of the user-defined container is changed to point to a port on the sidecar.
> 2) The readinessProbe of the sidecar container acts as a proxy to the user-defined container's port.

Can you help me understand?

@SergeyKanzhelev my preference order for this design point is:

1) Use readiness probes (and maybe ensure that the container is "ready" after startupProbe succeeds - no need to wait for both)
2) Use startup probes for this and document it well
3) use postStart for this and make that more visible - tell me that my pod is "starting" in the API

What do you think?

Author: This was discussed at today's Sidecar WG meeting. The concern of not being able to express "not ready, but started" was dominant. @smarterclayton mentioned other prior KEPs trying to take dependency on Ready state that wasn't accepted.

There was agreement on using the Startup probes to indicate this status. @matthyx offered to craft the proposal on [marking Pods as started](https://kubernetes.slack.com/archives/C0BP8PW9G/p1675194507712149) in general and applying it to sidecars once implemented.

As with every decision we keep the door open to change something after Alpha's feedback.

Reviewer 7: Sorry, I've been at a conference.

> > but we replace the container's declared probes with a probe of the sidecar
>
> I can interpret this 2 ways, which one is right:

> 1. The readinessProbe field of the user-defined container is changed to point to a port on the sidecar.
>1. The readinessProbe of the sidecar container acts as a proxy to the user-defined container's port.

It's actually (3):

The `readinessProbe` on the user container is _removed_. The `queue-proxy` binary has a `readinessProbe` defined, whose status depends on the results of the `queue-proxy` health-checking the user's container.

This allows us to implement a "retry quickly until container is started, then reduce pace to configured frequency" policy to improve initial readiness time without increasing steady-state load.

denote sections that are under active debate as follows:

```
<<[UNRESOLVED optional short context or usernames ]>>
Stuff that is being argued.
<<[/UNRESOLVED]>>
```

When editing KEPS, aim for tightly-scoped, single-topic PRs to keep discussions
focused. If you disagree with what is already in a document, open a new PR
with suggested changes.

One KEP corresponds to one "feature" or "enhancement" for its whole lifecycle.
You do not need a new KEP to move from beta to GA, for example. If
new details emerge that belong in the KEP, edit the KEP. Once a feature has become
"implemented", major changes should get new KEPs.

The canonical place for the latest set of instructions (and the likely source
of this file) is [here](/keps/NNNN-kep-template/README.md).

**Note:** Any PRs to move a KEP to `implementable`, or significant changes once
it is marked `implementable`, must be approved by each of the KEP approvers.
If none of those approvers are still appropriate, then changes to that list
should be approved by the remaining approvers and/or the owning SIG (or
SIG Architecture for cross-cutting KEPs).
-->
# KEP-NNNN: Your short, descriptive title

> <!--

Author: done

comparing to current approach. This can be eliminated by slight redesign of a
sidecars.

## Design Details

### Backward compatibility

The new field means that any Pod that is not using this field will behave the
same way as before.

The new field will only work with the proper control plane and kubelet. Upgrade
and downgrade scenarios will be covered in the further sections.

Outside of Kubernetes-controlled code, there might be third party controllers or
existing containers relying on the current behavior of Init containers.
Behaviors they can rely on:

- Assuming nothing is running in the Pod when Init container starts. For
  instance taking PID of the process assuming there are no sidecars.
- Incorrectly calculating the resource usage of a Pod not taking the new type of
  containers into account (e.g. some grafana dashboards may need modification)
- Stripping the restartPolicy:Always from Init container as an unknown field
  rendering Pod unfunctional as it will not pass the Init stage after this
- OPA rules may reject the Pod with new unknown fields because of failure to
  parse the new field

These potential incompatibilities will be documented.

> ### Resources calculation for scheduling and pod admission

Reviewer 8: Might be worth making it more explicit that we would have to enhance the [NodeResourcesFit scheduler plugin](https://github.com/kubernetes/kubernetes/blob/release-1.26/pkg/scheduler/framework/plugins/noderesources/fit.go#L160-L176) and take sidecar container resource request into consideration?

Author: added

When calculating whether Pod will fit the Node, resource limits and requests are
being calculated.

Resources calculation will change for Pod with sidecar containers. Today
resources are calculated as a maximum between the maximum use of an
InitContainer and Sum of all regular containers:

`Max ( Max(initContainers), Sum(Containers) )`

With the sidecar containers the formula will change.

Easiest formula would be to assume they are running the duration of the init
stage as well as regular containers.

`Max ( Max(nonSidecarInitContainers) + Sum(Sidecar Containers), Sum(Sidecar
Containers) + Sum(Containers) )`

However the true calculations will be different as all Init containers that
completed before the first sidecar containers will not need to account for any
sidecar containers for the maximum value calculation.

So the formula, assuming the function:

```
InitContainerUse(i) = Sum(sidecar containers with index < i) + InitContainer(i)
```

Is this:

```
Max ( Max( each InitContainerUse ) , Sum(Sidecar Containers) + Sum(Containers) ) 
```

> 

Author: I don't think the method will change, but we definitely need to test it. Updated text, thank you!

The feedback from SIG Scheduling was that this formulay needs to be implemented
in common library between scheduler and kubelet to avoid misalignment.

***Note:*** this formula is quite complex, maybe the result of this computation
can be exposed as a metric for external consumption.

### Topology and CPU managers

Preliminary code analysis didn't expose any issues introducing sidecar
containers. The biggest question is resources reuse for sidecar containers and
other Init containers, especially in cases of single NUMA node requirements and
such. This may be non-trivial. The decision on this is not blocking the KEP
though.

> TBD: add more details on resource reuse by sidecar containers

Author: thank you!

this scenario during startup (as it is probably the change you need to do to
handle sidecar crashes), for shutdown this approach won't work.

On shutdown the ordering behavior is arguably more important: if the logging
container is stopped first, logs for other containers are lost. No matter if
those containers queue them and retry to send them to the logging container, or
if they are persisted to a shared volume. The logging container is already
killed and will not be restarted, as the pod is shutting down. In these cases,
logs are lost.

The same things regarding startup and shutdown apply for a metrics container.

### Problems: service mesh

Service mesh presents a similar problem: you want the service mesh container to
be running and ready before other containers start, so that any inbound/outbound
connections that a container can initiate go through the service mesh.

A similar problem happens for shutdown: if the service mesh container is
terminated prior to the other containers, outgoing traffic from other apps will
be blackholed or not use the service mesh.

However, as none of these are possible to guarantee, most service meshes (like

> Linkerd and Istio), need to do several hacks to have the basic functionality.

Author: ```suggestion
Linkerd and Istio), use fragile and platform specific workarounds to have the basic functionality.
```

- OPA rules may reject the Pod with new unknown fields because of failure to
  parse the new field

These potential incompatibilities will be documented.

### Resources calculation for scheduling and pod admission

When calculating whether Pod will fit the Node, resource limits and requests are
being calculated.

Resources calculation will change for Pod with sidecar containers. Today
resources are calculated as a maximum between the maximum use of an
InitContainer and Sum of all regular containers:

`Max ( Max(initContainers), Sum(Containers) )`

With the sidecar containers the formula will change.

Easiest formula would be to assume they are running the duration of the init
stage as well as regular containers.

`Max ( Max(nonSidecarInitContainers) + Sum(Sidecar Containers), Sum(Sidecar
Containers) + Sum(Containers) )`

However the true calculations will be different as all Init containers that
completed before the first sidecar containers will not need to account for any
sidecar containers for the maximum value calculation.

So the formula, assuming the function:

```
InitContainerUse(i) = Sum(sidecar containers with index < i) + InitContainer(i)
```

Is this:

```

> Max ( Max( each InitContainerUse ) , Sum(Sidecar Containers) + Sum(Containers) ) 

Reviewer 9: I think [Pod overhead ](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-overhead/) should also be added to all of these calculations, not a big deal in this doc but another easy thing for re-implementors of this calculation to get wrong.

Author: added, thanks

together with explanation why this is acceptable.
-->

<!--
Additionally, for Alpha try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>
The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

- `<package>`: `<date>` - `<test coverage>`

##### Integration tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- <test>: <link to test coverage>

##### e2e tests

##### Pod lifecycle scenarios without sidecar containers

TBD: describe test cases that can be affected by introducing sidecar containers.

##### Pod lifecycle scenarios with sidecar containers

> TBD: describe test cases to test with sidecar containers.

Reviewer 9: I saw this section is TBD, but so it won't get missed, I think sidecar containers will affect LimitRange admission.  Otherwise a sidecar could be used to bypass a LimitRange.

Author: added

several use cases and force weird work-arounds in the applications.

The next sections expand on what the current problems are. But, to give more
context, it is important to highlight that some companies are already using a
fork of Kubernetes with this sidecar functionality added (not all
implementations are the same, but more than one company has a fork for this).

### Problems: jobs with sidecar containers

Imagine you have a Job with two containers: one which does the main processing
of the job and the other is just a sidecar facilitating it. This sidecar could
be a service mesh, a metrics gathering statsd server, etc.

When the main processing finishes, the pod won't terminate until the sidecar
container finishes too. This is problematic for sidecar containers that run
continuously.

There is no simple way to handle this on Kubernetes today. There are
work-arounds for this problem, most of them consist of some form of coupling
between the containers to add some logic where a container that finishes
communicates it so other containers can react. But it gets tricky when you have
more than one sidecar container or want to auto-inject sidecars.

> The sidecar will also not be restarted when it was OOM killed, which may render

Author: in case of a Job (restartPolicy:never) it will not. Let me clarify

the pod unusable if the sidecar provided secure communication to other
containers. 

### Problems: logging/metrics sidecar

A logging sidecar should start before several other containers, to not lose logs
from the startup of other applications. Let's call "main" container the app that
will log and "logging" container the sidecar that will facilitate it.

> If the logging container starts after the main app, some logs can be lost.

Author: Logging is still an important scenario. We do not guarantee logs will be delivered, but making it significantly easier to implement in majority of cases. I guess there are gradations: from reliable logs collection to easy to implemet startup logs upload (that is 80% reliable) to no startup logs or hard to implement startup logs.

Would moving this scenario to "use cases" section instead of Motivation be better?

> If the logging container starts after the main app, some logs can be lost.

Reviewer 2: I think I would just acknowledge that this is not a complete solution to the problem of sidecar reliability.

> If the logging container starts after the main app, some logs can be lost.

Reviewer 5: I think this is probably c&p from the old KEP I wrote. I think this sentence is not so important, it was to make clearer the next example (the next sentence).

> If the logging container starts after the main app, some logs can be lost.

Author: it is a copy paste indeed =) I chenged wording a bit to remove the "lost" language that implies reliability improvements

changes and perform the updates and push to the main container. This requires
two separate code paths today. Perhaps the same sidecar could be used for both
cases.

### Goals

This proposal aims to:
- make containers implementing the sidecar pattern first class citizens inside a
  Pod
- solve a Job completion issue when sidecars should run continuously
- allow mixing initContainers and sidecars for a choreographed startup sequence
- allow easy injection of sidecar containers in any Pod

### Non-Goals

This proposal doesn't aim to:
- support arbitrary dependencies graphs between containers
- act as a security control to enforce that pod containers only run while the
  sidecar is healthy. Restart of sidecar containers is a best effort
- allow to enforce security boundaries to sidecar containers different that
  other containers. For example, allow Istio to run privileged to configure ip
  tables and disable other containers from doing so
- (alpha) support containers termination ordering

## Proposal

The proposal is to introduce a `restartPolicy` field to Init containers and use
it to indicate that an initContainer is a sidecar container. The only supported

> value proposed in this KEP is `Always`. Kubelet will start these containers in

Author: they will follow the default behavior stated below, there will be no way to override it using this field

> value proposed in this KEP is `Always`. Kubelet will start these containers in

Reviewer 2: You say "The only supported value ... is Always" and "Other values for `restartPolicy` of containers will not be accepted".

I want you to spell out the default.  What happens to this field if I do not specifiy a value?  What default value does it get?

> value proposed in this KEP is `Always`. Kubelet will start these containers in

Reviewer 2: I'm still looking for a statement of what value this field gets if it is not specified

> value proposed in this KEP is `Always`. Kubelet will start these containers in

Author: next paragraph saying what if not specified.

the order with other Init containers, but instead of waiting for its completion,

> it will wait for the container startup completion (startup probe succeeded or if

Author: readiness is needed to indicate the whole pod readiness for load balancers in case of service mesh scenarios

> it will wait for the container startup completion (startup probe succeeded or if

Reviewer 2: That doesn't answer my question?  The container can be ready without the pod being ready, right?

If a container has a `startupProbe`, then I can see arguing that it should be "ready" as soon as that passes (we don't document that, but we should).

If a container DOESN'T have a `startupProbe`, shouldn't we wait for it to become ready before moving on to the next initContainer?

> it will wait for the container startup completion (startup probe succeeded or if

Reviewer 2: This still doesn't explain why `.status.containerStatuses[x].ready` is not the correct signal?

Defining this conditionally is confusing, but "ready" is pretty easy to define (or at least is a definition used elsewhere).

> it will wait for the container startup completion (startup probe succeeded or if

Author: this is being discussed above in the comment https://github.com/kubernetes/enhancements/pull/3761#discussion_r1085974934. Resolving this one

no startup probe defined, postStart completed).

These containers will not block Pod completion - if all regular containers
complete, sidecar containers will be terminated.

During the sidecar startup stage it will behave similarly to Init containers. If
Pod `restartPolicy` is `Never`, sidecar container failed during startup will NOT
be restarted. If Pod `restartPolicy` is `Always` or `OnFailure`, it will be

> restarted.

Author: added

Once sidecar container is started (postStart completed and startup probe
succeeded), this containers will be restarted even when the Pod `restartPolicy`
is `Never` or `OnFailure`. Furthermore, sidecar containers will be restarted
even during Pod termination.

In order to minimize OOM kills of sidecar containers, the OOM adjustment for
these containers will match or exceed the OOM score adjustment of regular
containers in the Pod. This intent to solve the issue
https://github.com/kubernetes/kubernetes/issues/111356 

As part of this KEP we also will be enabling for sidecar containers:
- PreStart and PostStop lifecycle handlers for sidecar containers
- All probes (startup, readiness, liveness)
- Readiness probes of sidecars will be used to determine the whole Pod
  readiness, but only after the main containers initialization started.
  Readiness probes will not start running before the Pod initialization stage
  completed 

```yaml
kind: Pod
spec:
  initContainers:
  - name: vault-agent
    image: hashicorp/vault:1.12.1

>     restartPolicy: OnFailure

Author: fixed. Leftover from past

  - name: istio-proxy
    image: istio/proxyv2:1.16.0
    args: ["proxy", "sidecar"]
    restartPolicy: Always
  containers:
  ...
```

### Future use of `restartPolicy` field

> In the Alpha version, no other values beyond Always will be acceptable. Going

Author: I agree. This section is written with some assumptions after thinking about it for a while. I think the main message here is that this field may make sense on other containers going forward.

How detailed do you see this message outlined? I really don't want to get into arguments on how this will work on other containers beyond "it will make sense for other containers/values" in future.

> In the Alpha version, no other values beyond Always will be acceptable. Going

Reviewer 2: I'd move ALL the future stuff to a later section and then you can include details and not have it be so distracting :)

> In the Alpha version, no other values beyond Always will be acceptable. Going

Author: Moved all future stuff to later section

being calculated.

Resources calculation will change for Pod with sidecar containers. Today
resources are calculated as a maximum between the maximum use of an
InitContainer and Sum of all regular containers:

`Max ( Max(initContainers), Sum(Containers) )`

With the sidecar containers the formula will change.

Easiest formula would be to assume they are running the duration of the init
stage as well as regular containers.

`Max ( Max(nonSidecarInitContainers) + Sum(Sidecar Containers), Sum(Sidecar
Containers) + Sum(Containers) )`

However the true calculations will be different as all Init containers that
completed before the first sidecar containers will not need to account for any
sidecar containers for the maximum value calculation.

So the formula, assuming the function:

```
InitContainerUse(i) = Sum(sidecar containers with index < i) + InitContainer(i)
```

Is this:

```
Max ( Max( each InitContainerUse ) , Sum(Sidecar Containers) + Sum(Containers) ) 
```

The feedback from SIG Scheduling was that this formulay needs to be implemented
in common library between scheduler and kubelet to avoid misalignment.

> ***Note:*** this formula is quite complex, maybe the result of this computation

Reviewer 2: "Metric" or "status" ?  It's not going to change in real-time right?

Author: no, not going to change in real time. Ideally we need a readonly calculated field on a Pod object I think. What do you mean by Status?

Reviewer 2: put it in pod status, not a metric :)

Author: @tzneal offered help writing down the proposal with more details taking also into account the inplace update KEP: https://github.com/kubernetes/enhancements/issues/1287

Author: draft proposal from @tzneal: https://docs.google.com/document/d/16_jl7_12-S6Nx-EYqWvtOeXZrRCJr4dboQBE5qewIwY/edit

Author: cc: @vinaykul please comment on this proposal

Reviewer 10: please review my thoughts here: https://github.com/kubernetes/enhancements/pull/3761#discussion_r1092020710

Author: resolving this comment since the discussion is in other comment

implemented to gather logs from the main app. Furthermore, if the logging
container is not yet started and the main app crashes on startup, those logs are
more likely to be lost (depends if logs go to a shared volume or over the
network on localhost, etc.). While you can modify your application to handle
this scenario during startup (as it is probably the change you need to do to
handle sidecar crashes), for shutdown this approach won't work.

On shutdown the ordering behavior is arguably more important: if the logging
container is stopped first, logs for other containers are lost. No matter if
those containers queue them and retry to send them to the logging container, or
if they are persisted to a shared volume. The logging container is already
killed and will not be restarted, as the pod is shutting down. In these cases,
logs are lost.

The same things regarding startup and shutdown apply for a metrics container.

### Problems: service mesh

Service mesh presents a similar problem: you want the service mesh container to
be running and ready before other containers start, so that any inbound/outbound
connections that a container can initiate go through the service mesh.

> A similar problem happens for shutdown: if the service mesh container is

Reviewer 7: Might be worth acknowledging the chicken/egg here between log collector exporting final logs and mesh logging last moments of shutdown, even if the solution is out of scope.

Author: added

- allow to enforce security boundaries to sidecar containers different that
  other containers. For example, allow Istio to run privileged to configure ip
  tables and disable other containers from doing so
- (alpha) support containers termination ordering

## Proposal

The proposal is to introduce a `restartPolicy` field to Init containers and use
it to indicate that an Init container is a sidecar container. Kubelet will start Init containers with
`restartPolicy=Always` in the order with other Init containers, but instead of
waiting for its completion, it will wait for the container startup completion.

The condition for startup completion is open question. It will be either:
- pod is ready (readiness probe succeeded) or
- startup probe succeeded or if no startup probe defined, `postStart` handler
completed. The decision on specific signal for container startup completion will
be locked on before implementation will start. See the section
["Pod startup completed condition"](#pod-startup-completed-condition)

As part of the KEP, init containers and regular containers will be split into
two different types. The field `restartPolicy` will only be introduced on init

> containers. The only supported value proposed in this KEP is `Always`. No other

Reviewer 7: Would it be clearer to define this as having the values `""` and `Always`, rather than introducing a special null-able state?

Author: null value seems to be easier for rollout and rollback. nullable field is plan of record as of now

values will be defined as part of this KEP. Moreover, the field will be

> nullable so the default value will be "no value". 

Author: I have added it to the Beta Graduation requirements

and more details can be found in the section ["Future use of `restartPolicy` field"]((#future-use-of--field))).

Sidecar containers will not block Pod completion - if all regular containers
complete, sidecar containers will be terminated.

During the sidecar startup stage the restart behavior will be similar to Init
containers. If Pod `restartPolicy` is `Never`, sidecar container failed during
startup will NOT be restarted and the whole Pod will fail. If Pod
`restartPolicy` is `Always` or `OnFailure`, it will be restarted.

Once sidecar container is started (`postStart` completed and startup probe
succeeded), this containers will be restarted even when the Pod `restartPolicy`
is `Never` or `OnFailure`. Furthermore, sidecar containers will be restarted
even during Pod termination.

In order to minimize OOM kills of sidecar containers, the OOM adjustment for
these containers will match or exceed the OOM score adjustment of regular
containers in the Pod. This intent to solve the issue
https://github.com/kubernetes/kubernetes/issues/111356 

As part of this KEP we also will be enabling for sidecar containers (those will
not be enabled for other Init containers):

> - PreStart and PostStop lifecycle handlers for sidecar containers

Author: fixed

- All probes (startup, readiness, liveness)

> - Readiness probes of sidecars will be used to determine the whole Pod

Reviewer 7: "will contribute to"?

Author: fixed

  and containers, but semantics that are mostly the same.
- Less 
  
Cons:
- There are slight differences between the semantics of `Always` for containers
  and `Always` for pods. The main difference is that a `initContainer` with
  `Always` will be terminated when the pod terminates and has no influence over
  the pod lifecycle. Also, for Pods, the default `restartPolicy` is
  [documented](https://github.com/kubernetes/kubernetes/blob/280473ebc4e45f9001f5f9789c318ff7329bc5f0/staging/src/k8s.io/api/core/v1/types.go#L2753-L2764)
  as `Always` but for `initContainers` it will default to `OnFailure`. We
  believe this can be addressed in the documentation of the existing enum
  fields.

When in future we may support `Always` on regular containers, there will be
interesting case of `Always` having a meaning of non-blocking container for the
Pod with `restartPolicy == Never` or `OnFailure`. This is easy to explain - Pod
lifecycle is controlled by containers with the matching restartPolicy. But it
may be slightly confusing and needs to be carefully reviewed at the time this
feature will be considered.

  
#### Alternatives

> - Introduce a new collection `infrastructureContainers` that replaces

Reviewer 7: I don't like it, but you could also support some sort of insertion or "placeholder value" in `initContainers` to help sequence these sidecar containers.

I only mention it in case it inspires someone else.

Author: added note about this alternative

  the semantic of this new value.
- Introduce a new field: `ambient: true` on all containers. This property will
  make containers be restarted all the time, and will not block the pod
  completion.
  - Pros: detaching sidecar proposal from the per-container `restartPolicy`
    proposal
  - Cons: this field is a new concept that will be introduced for the same
    property that is typically controlled by `restartPolicyß`.

### User Stories (Optional)

User stories are covered in Motivation section.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

The following is the list of hypothetical scenarios how users may decide to
abuse the feature or use it in a non-designed way. This is an exercise to
understand potential side effects from implementing the feature. We are looking
for side effects like abuse of the functionality to implement something
error-prone that sidecars were not designed for. Or causing issues like noisy

> neighbors, etc.

Reviewer 7: Is there a scenario here for ordering problems when attempting to inject two sidecars?  I haven't carefully thought through the mutating webhook semantics, but I seem to recall that the mutations are collected in parallel; I'm not sure how two injectors might agree on an order of container injection.

Author: Added scenario 1.a:

> ### Scenario 1.a
> 
>User decides to run regular container as a sidecar to simply guarantee the
> startup sequence of the containers. Users can already implement the startup sequencing by using blocking nature > of `postStart` hook. So using the sidecar
> containers instead is not adding any benefits.
>
> As for injecting sidecars into Pods with regular containers run as sidecars - 
> webhooks will likely inject sidecar containers to be first, so the risk is minimal here as well.

I cannot formulate to myself the risk related to different ordering of sidecars injection by the webhook

Reviewer 7: I suppose the risk is no different than trying to inject `initContainers` without sidecars.

logic inside with timeout or waiting for some cancellation signal from custom
job orchestration portal. This way users may implement execution with a time
limit defined by that logic. This is something that users cannot express today
“declaratively” and will be possible using new sidecar containers. One may
imagine a scenario when a third party jobs orchestration tool converts all
containers to sidecar containers programmatically and injects a single job
orchestration container into the Pod. This can be seen as an abuse of the
concept, and may lead to issues when pods terminate unexpectedly or Jobs meant
to be run as regular containers start being run as sidecars being restarted on
Completion. However we do not see a major problem with it as this use of sidecar
containers will unlikely lead to "unexpected" behavior - all side effects are
quite clear.

Another reason for users to implement regular containers as sidecars would be to
use any special properties kubelet may apply to sidecar containers. For example,
if restart backoff timeouts will be minimized for sidecar containers (which was
proposed and rejected), users may decide to use this feature for regular

> containers by running those as sidecars. Current proposal as written doesn’t

Reviewer 7: (except the OOM score adjustment)

Author: added clarification

start in order, sidecar resource usage needs to be summed into those init
containers that start after the sidecar.  Defining `InitContainerUse` as:

```
InitContainerUse(i) = Sum(sidecar containers with index < i) + Max(Spec.InitContainers[i].Resources, Status.InitContainerStatuses[i].ResourcesAllocated)
```

allows representing the new formula for a pods resource usage

```
Max ( Max( each InitContainerUse ) , Sum(Sidecar Containers) + Sum(each ContainerUse) ) + pod overhead
```

Even now, users still sometimes find how a pods effective resource requirements
are calculated confusing or are just unaware of the formula.  The mitigating
quality to this is that init container resource requests are usually lower than
the sum of non-init container resource requests, and can be ignored by users in
those cases.  Software that requires accurate pod resource requirement
information (e.g. kube-scheduler, kubelet, autoscalers) don't have that luxury.
Its too much to ask of users to perform this even more complex calculation
simply to know the amount of free capacity they need for a given resource to
allow a pod to schedule.

> #### Goals of exposing the Pod.ResourceAllocated field

Reviewer 10: As for adding a new field `PodStatus.ResourcesAllocated`, I don't favor it as currently proposed in this KEP. The ResourcesAllocated field is used to describe the resource levels at which each container in the pod has been admitted by a node (a.k.a achieved successful `canAdmitPod(...)`). This is not applicable to unassigned or Pending pods.

IIRC, pending pods that are not scheduled due to lack of resources have their pod conditions updated with the underlying reasons. I'm also struggling to see how adding this field equips users with something actionable (programmatically) that can justify its presence in the pod api.

I do think it is a great idea to expose a few simple helpers that computes a pod's resource requirements and allocations to be consumed by metrics, kubectl describe pod etc, and eliminate any code duplications.

> #### Goals of exposing the Pod.ResourceAllocated field

Reviewer 2: We did consider adding a whole new collection, but decided after much debate that it didn't add enough value to carry it's own weight.

Regarding `ResourcesAllocated`, do you object to the name or idea?  E.g. if it was `ResourcesRequested` ?

> #### Goals of exposing the Pod.ResourceAllocated field

Author: > Regarding `ResourcesAllocated`, do you object to the name or idea? E.g. if it was `ResourcesRequested` ?

We discussed this on today's Sidecar meeting. @smarterclayton had initially concern with introducing this field as there is already a metric for this exposed by scheduler.

But after taking thru scenarios:
- User wants to run `kubectl describe pod` to understand how big of the Node to allocate to schedule a pending pod
- User wants to see a real up to date allocated by the Pod, especially when the new requests are being applied using InPlace Pod Vertical Scaling (metrics wouldn't work for this as they show the "final" result).

we seems to be on the same page regarding the need to have a way to get this data quickly.

@vinaykul suggested maybe we add something specifically for `kubectl describe` command instead of changing the object. But I'm not sure if there are precedence.

One question is whether it is a field on PodStatus or a sub-resource of sort. @thockin I do not see any issues with having the field. Do you see any? Would it be better to have some sort of sub-resource?

> #### Goals of exposing the Pod.ResourceAllocated field

Reviewer 2: Subresources have a lot of overhead - I don't see how that's justified for this. This doesn't warrant an abstraction (eg /scale) and doesn't need RBAC isolation (eg /status).

I think a field seems just right

> #### Goals of exposing the Pod.ResourceAllocated field

Reviewer 10: > We did consider adding a whole new collection, but decided after much debate that it didn't add enough value to carry it's own weight.

I figured as much. To me, at first glance, a separate sidecar/support containers w/ resource allocations similar to normal containers + superseding restartPolicy {Always, OnFailure, Never} with Init containers -> Sidecar/Support containers -> normal/app containers flow seemed to address the service mesh user story when I walked through it. I know I don't have all the context, but I just had to ask :)

> 
> Regarding `ResourcesAllocated`, do you object to the name or idea? E.g. if it was `ResourcesRequested` ?

I don't favor the name, I was _maybe_ on the idea.

If we end up implementing [pod-level requests+limits KEP](https://github.com/kubernetes/enhancements/pull/1592) for aggregated sidecar container resource management, I want to reserve `ResourcesAllocated` for consistency/congruency with container-level `ResourcesAllocated` [as defined in my KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#api-changes) so that we can extend in-place resize, if desired, to pod level without causing confusion.

As for the idea, after attending today's sidecar meeting, I have a better handle on the need.  IIUC:
  - To handle scheduling failure due to resource paucity, we need to know say TotalResourceRequirements = max(max(initContainer), sum(Containers)) + podOverhead. (_Resize is not in play_)
  - Metrics needs TotalResourceAllocations = sum(max(Container.Requests, ContainerStatus.ResourcesAllocated))) + podOverhead (_Resize may be in play, scheduler assists InProgress resize_)

If subresource is too heavy for this, kubectl describe / extended PodScheduled condition message could provide this info. But it is not a great choice for programmatic processing where we should respond with objects.

> #### Goals of exposing the Pod.ResourceAllocated field

Reviewer 10: @thockin Adding a field `PodSpec.TotalResourcesRequested` seems to address the needs of both scheduling failure and metrics. Let's hold `ResourcesAllocated` in reserve for a possible pod-level requests/limits feature. WDYT?

> #### Goals of exposing the Pod.ResourceAllocated field

Reviewer 2: Why spec and not status?  I think spec is a request, which is the pod-level requests+limits KEP, but what we're trying to represent here is the result of  a non-obvious calculation, to save people from needing to do it themselves.

> #### Goals of exposing the Pod.ResourceAllocated field

Reviewer 10: Oops! Because I have buttons for eyes 😵‍💫 :) 
It should have been [PodStatus.TotalResourcesRequested](https://github.com/kubernetes/enhancements/pull/3761#discussion_r1093297192) which imho more accurately represents what's needed for the use case of adding nodes to get "insufficient resources/unschedulable" pods going as well as what is being proposed for metrics. I agree we should add it .. other options are not great. Thoughts on my suggested name?

> #### Goals of exposing the Pod.ResourceAllocated field

Reviewer 2: LOL.  I don't object to the name, but we all know how good I am at naming things (cough)Ingress.

> #### Goals of exposing the Pod.ResourceAllocated field

Author: Resolving for now. Name can be decided during the API review of a specific PR

- allow mixing initContainers and sidecars for a choreographed startup sequence
- allow easy injection of sidecar containers in any Pod
- (to be evaluated after alpha) allow to implement sidecar containers that will
  guaranteed to be running longer than regular containers 

### Non-Goals

This proposal doesn't aim to:
- support arbitrary dependencies graphs between containers
- act as a security control to enforce that pod containers only run while the
  sidecar is healthy. Restart of sidecar containers is a best effort
- allow to enforce security boundaries to sidecar containers different that
  other containers. For example, allow Istio to run privileged to configure ip
  tables and disable other containers from doing so
- (alpha) support containers termination ordering

## Proposal

The proposal is to introduce a `restartPolicy` field to Init containers and use
it to indicate that an Init container is a sidecar container. Kubelet will start Init containers with
`restartPolicy=Always` in the order with other Init containers, but instead of
waiting for its completion, it will wait for the container startup completion.

> The condition for startup completion is open question. It is UNRESOLVED now

Author: this was resolved

  completed, or
- `postStart` handler completed.

The decision on specific signal for container startup completion will
be locked on before implementation will start. See the section
["Pod startup completed condition"](#pod-startup-completed-condition)

As part of the KEP, init containers and regular containers will be split into
two different types. The field `restartPolicy` will only be introduced on init
containers. The only supported value proposed in this KEP is `Always`. No other
values will be defined as part of this KEP. Moreover, the field will be
nullable so the default value will be "no value". 

Other values for `restartPolicy` of containers will not be accepted and
containers will follow the logic is currently implemented (documented
[here](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#:~:text=if%20a%20pod's%20init%20container%20fails%2C%20the%20kubelet%20repeatedly%20restarts%20that%20init%20container%20until%20it%20succeeds.%20however%2C%20if%20the%20pod%20has%20a%20restartpolicy%20of%20never%2C%20and%20an%20init%20container%20fails%20during%20startup%20of%20that%20pod%2C%20kubernetes%20treats%20the%20overall%20pod%20as%20failed.)

> and more details can be found in the section ["Future use of `restartPolicy` field"]((#future-use-of--field))).

Author: fixed

Sidecar containers will not block Pod completion - if all regular containers
complete, sidecar containers will be terminated.

During the sidecar startup stage the restart behavior will be similar to Init
containers. If Pod `restartPolicy` is `Never`, sidecar container failed during
startup will NOT be restarted and the whole Pod will fail. If Pod

> `restartPolicy` is `Always` or `OnFailure`, it will be restarted.

Reviewer 10: On the topic of per-container restartPolicy, in the last SIG-node meeting, I had wondered if we should bring the newly added per-container resizePolicy.restartPolicy under the restartPolicy proposed here. On further consideration, it is best not to conflate them. Per-container resizePolicy influences the outcome of ContainersToStart & ContainersToKill in `computePodActions` during resize and can operate orthogonal to per-container restartPolicy with the same validation constraints as defined today.

And for simplicity, we can perhaps build it such that if per-container restartPolicy is specified, it supercedes the pod-level restartPolicy value, JMHO.

Reviewer 9: > As for adding a new field PodStatus.ResourcesAllocated, I don't favor it as currently proposed in this KEP. The ResourcesAllocated field is used to describe the resource levels at which each container in the pod has been admitted by a node (a.k.a achieved successful canAdmitPod(...)). This is not applicable to unassigned or Pending pods.

Once the pod is scheduled to the node, `PodStatus.ResourcesAllocated`, would be this value.  I can see the argument that its misusing the name "Allocated" prior to being scheduled, but I think the benefit of clearly showing the resources required for the pod to schedule outweighs that.

> IIRC, pending pods that are not scheduled due to lack of resources have their pod conditions updated with the underlying reasons. I'm also struggling to see how adding this field equips users with something actionable (programmatically) that can justify its presence in the pod api.

The condition is updated, however the message isn't helpful for determining how to solve the problem.  E.g. Below I know that I have insufficient CPU in this case, but how insufficient?  

```
  - lastProbeTime: null
    lastTransitionTime: "2023-01-31T15:23:24Z"
    message: '0/2 nodes are available: 1 Insufficient cpu, 1 node(s) had taint {CriticalAddonsOnly:
      true}, that the pod didn''t tolerate.'
    reason: Unschedulable
    status: "False"
    type: PodScheduled
```

If I wanted to determine how much more CPU the node would need so that this pod can run, I need to determine the resources required by the pod.  This requires the user to manually implement the multi-part calculation described in the KEP.

Reviewer 10: @tzneal If possible, I want to reserve the name `PodStatus.ResourcesAllocated` for a possible future pod-level requests+limits KEP. On further thought, neither kubectl describe nor extending pod condition message are good options. We should respond with objects. `PodStatus.TotalResourcesRequested` better represents what you need as well as what metrics needs. Please see comment: https://github.com/kubernetes/enhancements/pull/3761#discussion_r1092772324 . WDYT?

Reviewer 9: That seems fine to me, I chose `ResourcesAllocated` to avoid a proliferation of names even though its not precise, but if you suspect there's a need to use it for something else it makes sense to pick something better here.

Reviewer 10: Besides the possibility of the pod level requests+limits KEP, we have defined ResourcesAllocated in the [In-place pod resize KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#api-changes) as the amount of resources node has allocated for the pod. It is odd to overload the _Allocated_ term for pending pods, (no resources allocated yet) and for pods that desire resize but node has not granted (Infeasible, Deferred, Proposed resize statuses).

Suggestion: If it is fine with you, could you please work with @SergeyKanzhelev to add this particular use case to the 'Goals of exposing the Pod.ResourceAllocated field' section. Giving the ability to cluster autoscaler, karpenter etc to collect a bunch of resource unavailable/unschedulable pods and easily sum up their TotalResourcesRequested to add nodes is a very strong justification to have this field in PodStatus. I only ask that <begin shameless plug> you use in-place resize to right-size over-provisioned pods first to help keep carbon footprint low and save some $ while at it before adding nodes :) <end shameless plug>

Author: added word to word, thank you!

  make containers be restarted all the time, and will not block the pod
  completion.
  - Pros: detaching sidecar proposal from the per-container `restartPolicy`
    proposal
  - Cons: this field is a new concept that will be introduced for the same
    property that is typically controlled by `restartPolicyß`.

### User Stories (Optional)

User stories are covered in Motivation section.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

The following is the list of hypothetical scenarios how users may decide to
abuse the feature or use it in a non-designed way. This is an exercise to
understand potential side effects from implementing the feature. We are looking
for side effects like abuse of the functionality to implement something
error-prone that sidecars were not designed for. Or causing issues like noisy
neighbors, etc.

### Scenario 1. User decides to use sidecars as a way to run regular containers

> 

Author: fixed

and downgrade scenarios will be covered in the further sections.

Outside of Kubernetes-controlled code, there might be third party controllers or
existing containers relying on the current behavior of Init containers.
Behaviors they can rely on:

- Assuming nothing is running in the Pod when Init container starts. For
  instance taking PID of the process assuming there are no sidecars.
- Incorrectly calculating the resource usage of a Pod not taking the new type of
  containers into account (e.g. some grafana dashboards may need modification)
- Stripping the restartPolicy:Always from Init container as an unknown field
  rendering Pod unfunctional as it will not pass the Init stage after this
- OPA rules may reject the Pod with new unknown fields because of failure to
  parse the new field

These potential incompatibilities will be documented.

### Pod startup completed condition

As part of this KEP, we need to decide on a condition that Pod needs to meet to
be recognized as "started". This will be used to sequence Init containers
properly and give enough time for sidecar container to complete it's
initialization.

> ## Readiness probes

Author: fixed

sequentially with the `postStart` hook being a blocking call.

Using `postStart` has some debugging challenges in how logs are reported (they
aren't) and the status is displayed. If this option will be used, we should
consider improving this situation.

With `postStart` simple sidecars will start faster than if any probes are used.
However it brings additional challenges for sidecar authors as majority of logic
may be duplicated and needs to be authored, e.g. a shell binary + a loop script
to retry `curl`.

### Resources calculation for scheduling and pod admission

When calculating whether Pod will fit the Node, resource limits and requests are
being calculated.

Resources calculation will change for Pod with sidecar containers. Today
resources are calculated as a maximum between the maximum use of an
InitContainer and Sum of all regular containers:

`Max ( Max(initContainers), Sum(Containers) )`

With the sidecar containers the formula will change.

Easiest formula would be to assume they are running the duration of the init
stage as well as regular containers.

`Max ( Max(nonSidecarInitContainers) + Sum(Sidecar Containers), Sum(Sidecar
Containers) + Sum(Containers) )`

> However the true calculations will be different as all Init containers that

Author: done in most places I found

represent the actual resources in use.  The KEP notes that:

> Schedulers should use the larger of `Spec.Containers[i].Resources` and
> `Status.ContainerStatuses[i].ResourcesAllocated` when considering available
> space on a node.

We can introduce `ContainerUse` to represent this value:

```
ContainerUse(i) = Max(Spec.Containers[i].Resources, Status.ContainerStatuses[i].ResourcesAllocated)
```

In the absence of KEP 1287, or if the feature is disabled, `ContainerUse` is
simply:

```
ContainerUse(i) = Spec.Containers[i].Resources
```

The sidecar KEP also changes that calculation to be more complicated as sidecar
containers are init containers that do not terminate.  Since init containers
start in order, sidecar resource usage needs to be summed into those init
containers that start after the sidecar.  Defining `InitContainerUse` as:

```
InitContainerUse(i) = Sum(sidecar containers with index < i) + Max(Spec.InitContainers[i].Resources, Status.InitContainerStatuses[i].ResourcesAllocated)
```

allows representing the new formula for a pods resource usage

```
Max ( Max( each InitContainerUse ) , Sum(Sidecar Containers) + Sum(each ContainerUse) ) + pod overhead
```

> Even now, users still sometimes find how a pods effective resource requirements

Author: fixed

are calculated confusing or are just unaware of the formula.  The mitigating
quality to this is that init container resource requests are usually lower than
the sum of non-init container resource requests, and can be ignored by users in
those cases.  Software that requires accurate pod resource requirement
information (e.g. kube-scheduler, kubelet, autoscalers) don't have that luxury.

> Its too much to ask of users to perform this even more complex calculation

Author: fixed

- Provide a well documented, reusable, exported function to be used to calculate
  the effective resource requirements for a v1.Pod struct.
- Eliminate duplication of the pod resource requirement calculation within
  `kubelet` and `kube-scheduler`.

#### Implementation details

We propose making two changes to satisfy the two primary consumers of the
effective pod resource requests, users and other Kubernetes ecosystem software
components.

The first change is to add a field to `PodStatus` to represent the effective
resource requirements for the pod.  The field is named `ResourcesAllocated` which
matches the naming of the container status field from the KEP 1287. This field
allows users to inspect running and pending pods to quickly determine their
effective resource requirements. The field will be first populated by scheduler
in the [updatePod](https://github.com/kubernetes/kubernetes/blob/815e651f397a6f5691efb26b0d4f3bddb034668c/pkg/scheduler/schedule_one.go#L943). The updating of this field would occur the existing
[generateAPIPodStatus](https://github.com/kubernetes/kubernetes/blob/a668924cb60901b413abc1fe7817bc7969167064/pkg/kubelet/kubelet_pods.go#L1459) method.

```

> // ResourcesAllocated is the effective resource requirements for this pod, taking into consideration init containers, sidecars, containers and pod overhead.
> // More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/  
> // +optional  
> ResourcesAllocated ResourceRequirements `json:"resources,omitempty" protobuf:"bytes,8,opt,name=resources"`

Reviewer 11: ```suggestion
// ResourcesAllocated is the effective resource requirements for this pod, taking into consideration init containers, sidecars, containers and pod overhead.
// More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/  
// +optional  
ResourcesAllocated ResourceList `json:"resourcesAllocated,omitempty" protobuf:"bytes,8,opt,name=resourcesAllocated"`
```

Maybe we need to change the json field name and type?
- name: resources -> resourcesAllocated
- type: ResourceRequirements -> ResourceList ?

Reviewer 9: Yes, I changed the name a few times and missed the JSON tag.  I think a ResourceList makes sense as well, it also matches the container `ResourcesAllocated` field in the in-place resizing KEP.

Author: fixed

though.

From the code, it appears that init containers are treated exactly like
application containers so we don't need to change anything from resource
management point of view. I found references in the code where all the
containers (init containers and application containers) were coalesced before
resources (CPUs, memory and devices) are allocated to them. Here are a few
examples:

- Container Manager:
  https://github.com/kubernetes/kubernetes/blob/release-1.26/pkg/kubelet/cm/container_manager_linux.go#L708
- CPU Manager:
  https://github.com/kubernetes/kubernetes/blob/release-1.26/pkg/kubelet/cm/cpumanager/policy_static.go#L490
- Memory Manager:
  https://github.com/kubernetes/kubernetes/blob/release-1.26/pkg/kubelet/cm/memorymanager/policy_static.go#L372
- Topology Manager:
  - https://github.com/kubernetes/kubernetes/blob/release-1.26/pkg/kubelet/cm/topologymanager/scope.go#L137
  - https://github.com/kubernetes/kubernetes/blob/release-1.26/pkg/kubelet/cm/topologymanager/scope_container.go#L52
  - https://github.com/kubernetes/kubernetes/blob/release-1.26/pkg/kubelet/cm/topologymanager/scope_pod.go#L58

### Termination of containers

> Required for graduation to Beta.

Author: updated text

- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

### Upgrade / Downgrade Strategy

#### Upgrade strategy

Existing sidecars will still work as intended, as in the past we didn't
recognize this type of container.

#### Downgrade strategy

Due to the new field added to initContainers to turn them into sidecars,
downgrading to the version without this feature will not be possible -  the Pod
spec will be rejected by the control plane and all kubelets during
unmarshalling. Workloads will have to be deleted and recreated with the old way

> of handling sidecars.

Reviewer 12: Right, so then my sidecars will revert to initContainers and won't run as intended, as you suggest below.

The same problem will arise if the user wants to deploy such workloads on older

> clusters, or clusters without the feature flag enabled.

Reviewer 12: Right, we probably need to rephrase this part... I suspect we'll be using pointers for the new field to be able to differentiate between the 2 cases.

> clusters, or clusters without the feature flag enabled.

Author: fixed

### Version Skew Strategy

Version skew is possible between the control plane and worker nodes as both
should be aware of the new field used to flag sidecars inside initContainers.

Therefore all cluster nodes, including master nodes, must be upgraded before the
user can deploy sidecars using the new syntax.

Also, since the feature flag applies to both kubelet and the control plane,
similarly all cluster nodes need to have it enabled before deploying Pods with

> sidecars.

Reviewer 12: I don't see how we can make some magic here... I'm not planning on injecting these into `containers` to make sure they run always.

> sidecars.

Reviewer 12: Maybe it's safer to detect that and bail out to prevent disillusions?

> sidecars.

Author: updated text

for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `SidecarContainers`

>   - Components depending on the feature gate: kubelet, kube-apiserver

Author: fixed, aligned with kep.yaml

- [ ] Other
  - Describe the mechanism:
  - Will enabling / disabling the feature require downtime of the control
    plane?
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).

###### Does enabling the feature change any default behavior?

No.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> Yes. Pods that had sidecars will need to be deleted and recreated without them.

Author: Updated text.

I'm surprised by this statement in https://github.com/kubernetes/enhancements/pull/3298/files:

>Nodes do not currently support in-place upgrades, so pods will be recreated on
node upgrade and downgrade. No special handling or consideration is needed to
support this.

Not sure what in-place meant there, maybe the fact that upgrade without draining is not supported.

Also another KEP suggested at PRR meeting few weeks back as an example also doesn't have this well explained: https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/1205-bound-service-account-tokens/README.md#rollout-upgrade-and-rollback-planning

###### What happens if we reenable the feature if it was previously rolled back?

> Pods will have to be re-created to use the feature.

Reviewer 12: (by the way thanks for the interesting thoughts here)

> Pods will have to be re-created to use the feature.

Author: fixed

###### Are there any tests for feature enablement/disablement?

> Not required.

Reviewer 12: Right, but to be clear this scenario is only needed for going into beta... I don't think we expect being able to disable a feature in alpha without some issues.

> Not required.

Author: updated text

The proposal is to introduce a `restartPolicy` field to Init containers and use
it to indicate that an Init container is a sidecar container. Kubelet will start Init containers with
`restartPolicy=Always` in the order with other Init containers, but instead of
waiting for its completion, it will wait for the container startup completion.

> The condition for startup completion is open question. It will be either:

Reviewer 2: Should this be marked as UNRESOLVED ?

When will it be resolved?

Author: This decision is not blocking the overall KEP from my perspective. I want to get more feedback on it before finalizing. 

My understanding is that your position is `Ready` condition would be an easy and straightforward solution, but not strongly against relying on startup probes. Please correct me if I'm wrong

Reviewer 2: I meant that you should mark it as unresolved and indicate what / when we will resolve it.

That is my position, yes

Author: marked as unresolved
